### PR TITLE
Fix observe() method when called for SGVector/SGMatrix.

### PR DESCRIPTION
**Pull request by geektoni**
* Fix `observe()` method when called for `SGVector`/`SGMatrix`. The given `SGVector/SGMatrix` was not cloned correctly (since copying them just copy the pointer to the raw data and it does not make a clone of the data itself).

<del>I moved some useful type traits from `type_case.h` to `base_types.h` since I did not want to pollute too much `SGObject.h` with unnecessary headers. </del>

Any comments on this change? @karlnapf @vigsterkr @gf712

### CODE REVIEW

File: src/shogun/base/base_types.h [hunk 1]
```diff
@@ -68,6 +70,121 @@ namespace shogun
...
+		T_SGVECTOR_FLOATMAX = 18,
+		T_SGVECTOR_INT32 = 19,
+		T_SGVECTOR_INT64 = 20,
+		T_SGMATRIX_FLOAT32 = 21,
+		T_SGMATRIX_FLOAT64 = 22,
+		T_SGMATRIX_FLOATMAX = 23,
+		T_SGMATRIX_INT32 = 24,
+		T_SGMATRIX_INT64 = 25,
+		T_UNDEFINED = 26
+	};
```
Comment (gf712) [inline src/shogun/base/base_types.h]: Do you think it would be worth adding all the combinations of containers and primitive types, or some macro that would do that? Because right now stuff like `SGVector<bool>` isn't included (whether or not we actually use it is another question...)

File: src/shogun/base/base_types.h [hunk 2]
```diff
@@ -68,6 +70,121 @@ namespace shogun
...
+
+		template <typename T>
+		struct is_sg_vector : public std::false_type
+		{
+		};
+
+		template <typename T>
+		struct is_sg_matrix : public std::false_type
+		{
+		};
```
Comment (gf712) [inline src/shogun/base/base_types.h]: Remember that these structs do not check all possible `T`/`SGVector<T>`/`SGMatrix<T>`, this is to check if the type is registered at runtime to then use the dispatching. To check if a type is `SGVector<T>` you should do something like:
```
template <typename T>
struct is_sg_vector
{
    using type = std::false_type;
};
template <typename T>
struct is_sg_vector<SGVector<T>>
{
  using type = std::true_type;
};
```

Comment (gf712) [reply src/shogun/base/base_types.h]: (That's why these things were hidden in their own namespace)

File: src/shogun/base/base_types.h (hunk 1 above)
Comment (geektoni) [reply src/shogun/base/base_types.h]: I guess it would make sense if we were to use it somewhere in Shogun. For instance, if we find a machine which needs to emit `SGVector<bool>` then we will add it.

File: src/shogun/base/base_types.h (hunk 2 above)
Comment (geektoni) [reply src/shogun/base/base_types.h]: Argh you are indeed right. Those need to be left in their namespace. In the end, I just need to check if it is an SGVector/SGMatrix (an not if they match also with a type `T`).

File: src/shogun/base/SGObject.h [hunk 3]
```diff
@@ -1306,13 +1315,30 @@ void CSGObject::observe(
...
 {
 	auto param = this->get_parameter(BaseTag(name));
 	auto obs = some<ObservedValueTemplated<T>>(
 		step, name, any_cast<T>(param.get_value()), param.get_properties());
 	this->observe(obs);
 }
+
+template<class T,
+		typename std::enable_if_t<
+				is_sg_vector<T>::value ||
```
Comment (karlnapf) [inline src/shogun/base/SGObject.h]: what about std::vector?

File: src/shogun/base/SGObject.h [hunk 4]
```diff
@@ -1306,13 +1315,30 @@ void CSGObject::observe(
...
+				is_sg_matrix<T>::value>*>
+void CSGObject::observe(const int64_t step, const std::string& name) const
+{
+	auto param = this->get_parameter(BaseTag(name));
+	auto obs = some<ObservedValueTemplated<T>>(
```
Comment (karlnapf) [inline src/shogun/base/SGObject.h]: I also wonder: couldn't we do this inside the `ObservedValueTemplated` class?

File: src/shogun/base/SGObject.h (hunk 3 above)
Comment (karlnapf) [reply src/shogun/base/SGObject.h]: btw this is very similar to the problem we have in `any.h` with `clone` and `equals`. There we do the following: we check (in order) whether a type has `->clone`, then `.clone`, and if not we use the copy ctor. We also cover cases such as raw arrays and `std::vector`. I think an approach like this would be better here as well, as otherwise we will start patching and patching the template signature here, causing a change in `SGObject` every time....

Comment (geektoni) [reply src/shogun/base/SGObject.h]: Do you mean something like this https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/any.h#L276?

Comment (geektoni) [reply src/shogun/base/SGObject.h]: > what about std::vector?

`std::vector` has a copy constructor so there is no problem. The issue was with `SGVector` since its copy constructor does not copy also the memory, but it copies only the reference to the memory (hence we have multiple `SGVector`s that point to the same location).

File: src/shogun/base/SGObject.h (hunk 4 above)
Comment (geektoni) [reply src/shogun/base/SGObject.h]: I would prefer to keep a separate method to construct an `ObservedValueTemplated` object from a `Tag`.

Comment (geektoni) [reply src/shogun/base/SGObject.h]: If you are talking about a method to construct a new `ObservedValueTemplated` :)

File: src/shogun/base/SGObject.h (hunk 3 above)
Comment (geektoni) [reply src/shogun/base/SGObject.h]: @karlnapf have a look at this one: https://github.com/geektoni/shogun/commit/0a9a92e8012fc7352e4d757cb85f0681145854fa. This is basically the `any.h` code, but it should do exactly what we are looking for. Still WIP though. Feel free to criticize it :)

Comment (karlnapf) [reply src/shogun/base/SGObject.h]: Cool! :) That seems ok to me. The only critique I have is now we have the code twice within the codebase.... maybe you can make use of that in `any.h` too and delete it there? Or instead use the implementation from the `any_detail` namespace (since any is included in your file)
If not, also the filename can maybe just be `clone.h`

Comment (geektoni) [reply src/shogun/base/SGObject.h]: The `any.h` implementation uses some casts to `void*` and it also uses a `void** storage` argument to store the cloned value. These are things which are not exactly needed by our use case, so I thought it would have been better to have a separate `clone` which can be used from Shogun (and that is not `Any` specific).

Comment (karlnapf) [reply src/shogun/base/SGObject.h]: I see, seems reasonable ... :)

**Comment by geektoni** [comment]
@karlnapf any final comments before I merge this? :)

File: tests/unit/base/MockObject.h [hunk 5]
```diff
@@ -316,6 +316,13 @@ namespace shogun
 			watch_method("some_method", &CMockObject::some_method);
 		}
 
+		virtual CSGObject* create_empty() const
```
Comment (karlnapf) [inline tests/unit/base/MockObject.h]: curios: why did you add this here

**Comment by karlnapf** [comment]
good!

File: tests/unit/base/MockObject.h (hunk 5 above)
Comment (geektoni) [reply tests/unit/base/MockObject.h]: Basically, since `observe()` emit a clone of the given value, we happens to call `MockObject->clone()`. However, the standard `SGObject` implementation of `clone()` assumes that there is a corresponding `create()` method, for that element, inside `class_list.h`. This is not true for `MockObject` since it is used only for testing purpose. Therefore, I needed to override `create_empty` such to return a new instance without relying on the `class_list.h` methods.

Comment (karlnapf) [reply tests/unit/base/MockObject.h]: ah yep, I remember that stuff now. thanks!
